presto-main/src/main/java/com/facebook/presto/spiller/GenericPartitioningSpiller.java

```diff
@@ -0,0 +1,169 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.spiller;
+
+import com.facebook.presto.memory.AggregatedMemoryContext;
+import com.facebook.presto.operator.SpillContext;
+import com.facebook.presto.operator.exchange.LocalPartitionGenerator;
+import com.facebook.presto.spi.Page;
+import com.facebook.presto.spi.PageBuilder;
+import com.facebook.presto.spi.type.Type;
+import com.google.common.collect.ImmutableList;
+import com.google.common.io.Closer;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import it.unimi.dsi.fastutil.ints.IntArrayList;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+import java.io.IOException;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
+import java.util.function.IntPredicate;
+import java.util.function.Supplier;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.base.Verify.verify;
+import static io.airlift.concurrent.MoreFutures.getFutureValue;
+import static java.util.Objects.requireNonNull;
+
+@ThreadSafe
+public class GenericPartitioningSpiller
+        implements PartitioningSpiller
+{
+    private final List<Type> types;
+    private final int partitionsCount;
+    private final PageBuilder[] pageBuilders;
+    private final LocalPartitionGenerator partitionGenerator;
+    private final Supplier<SpillContext> spillContextSupplier; // TODO use. Why Supplier ?
+    private final SingleStreamSpiller[] spillers;
+    private final Closer closer = Closer.create();
+    private boolean readingStarted;
+    private Set<Integer> spilledPartitions = new HashSet<>();
+
+    public GenericPartitioningSpiller(
+            List<Type> types,
+            LocalPartitionGenerator partitionGenerator,
+            int partitionsCount,
+            Supplier<SpillContext> spillContextSupplier,
+            AggregatedMemoryContext memoryContext,
+            SingleStreamSpillerFactory spillerFactory)
+    {
+        this.partitionsCount = partitionsCount;
```

Comment:
Get number of partitions from `LocalPartitionGenerator`.
You need to extend its interface to expose that but it seems less error prone to have single source of number of partitions we are working with.